man/hfst-fst2txt.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH HFST-FST2TXT "1" "April 2026" "HFST" "User Commands"
.SH NAME
hfst-fst2txt \- =Print transducer in AT&T, dot, prolog or pckimmo format
.SH SYNOPSIS
.B hfst-fst2txt
[\fI\,OPTIONS\/\fR...] [\fI\,INFILE\/\fR]
.SH DESCRIPTION
Print transducer in AT&T, dot, prolog or pckimmo format
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.TP
\fB\-\-colour\fR[=\fI\,WHEN\/\fR]
Print in colour WHEN:
.TP
\fB\-\-color\fR[=\fI\,WHEN\/\fR]
always, never, auto (default)
.SS "Input/Output options:"
.TP
\fB\-i\fR, \fB\-\-input\fR=\fI\,INFILE\/\fR
Read input transducer from INFILE
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,OUTFILE\/\fR
Write output transducer to OUTFILE
.SS "Text format options:"
.TP
\fB\-w\fR, \fB\-\-print\-weights\fR
If weights are printed in all cases
.TP
\fB\-D\fR, \fB\-\-do\-not\-print\-weights\fR
If weights are not printed in any case
.TP
\fB\-f\fR, \fB\-\-format\fR=\fI\,TFMT\/\fR
Print output in TFMT format [default=att]
.PP
If OUTFILE or INFILE is missing or \-, standard streams will be used.
Unless explicitly requested with option \fB\-w\fR or \fB\-D\fR, weights are printed
if and only if the transducer is in weighted format.
TFMT is one of {att, dot, prolog, pckimmo}.
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://github.com/hfst/hfst/issues>
.PP
hfst\-fst2txt home page:
<https://github.com/hfst/hfst/wiki/HfstFst2Txt>
.br
General help using HFST software:
<https://github.com/hfst/hfst/wiki>
.SH COPYRIGHT
Copyright \(co 2017 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS for hfst-tools and library

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

## Noteworthy changes in 3.17.1

* fixes logic error where `hfst-lexc` would treat a warning about implicit
  alphabet in XREs as a critical error even when `-Werror` is not asked

## Noteworthy changes in 3.17.0

* API breaks: HfstTokenizer, LexcCompiler, `read_in_att_format` functions
  in HfstTransducer, HfstBasicTransducer, etc. to accommodate the following
  unicode handling changes and error reporting:
* Character "tokenisers" that used to operate on bytes or UTF-8 sequences will
  now operate on [graphemes](https://www.unicode.org/reports/tr29/tr29-23.html#Grapheme_Cluster_Boundaries),
  this means that by default arcs in FSAs are expected to match combining
  grapheme clusters, e.g. a base character plus all diacritics.
  The implementation is based on ICU implementation of `u_brk`
    * in lexc string entries, graphemes are compiled into a arcs
    * in xerox regex curly braces, as well
    * in tokenise without --multichar mode, input is tokenised by graphemens
    * etc. etc. tbd
* negative weights yield warnings now unless turned off
* tools will have more gcc-style warning control: -Wall -Werror -Wno-this-error
* lexc warnings are systematically treated as errors with -Werror now
* colourised more warnings and errors

## Noteworthy changes in 3.16.2

* Support for `>OpenFst-1.8.0`
    * only functional change is removal of n-best paths search for log weight
      automata, it probably has errored out before but has slightly different
      error now (HfstException instead of openfst exit)
* Optimisations to hfst-pmatch2fst alphabet handling

## Noteworthy changes in 3.16.1

* Bug fixes:

    * hfst-tokenize escapes both \r and \n now

    * regexp parsing more robust with comments and linebreaks

* Removal of some bundled libraries including foma and openfst, install
  suitable versions to compile support for them

* ICU stabilised as unicode handler over all

* experimental C API for rust bindings (unstable)

* new tools (testing):

    * hfst-check-alpha for checking and comparing alphabets in automata and
      their metadata

    * hfst-dump-alphabets for reviewing alphabets easier than hfst-summarize
      grepping

    * hfst-kill-paths for efficient path removal

* introduced clang-format to work towards standard code formatting

Noteworthy changes in 3.16.0
----------------------------

* Bugfixes:

  * During conversion between optimized-lookup and other formats, the alphabet
    was not being preserved, leading to incorrect behaviour in some cases.
    This has been fixed.

  * Fix segfaults in hfst-summarize, hfst-disjunct and foma back-end.

  * Force ICU, fixes #501.

  * Use unique_ptr for input/ouput stream memory allocation (#547).


* Improvements:

  * hfst-pmatch archives now no longer include or parse redundant alphabets,
    which in some cases can be a meaningful size and speed saving.

  * Multiple memory leaks fixed.

  * Drop support for TR1 standard assuming that C++14 is minimal requirement.

  * Drop legacy wrappers for foma and lexc (#532).

  * Use upstream Foma and by default upstream OpenFST.

  * Use C++17, or C++14 with bundled OpenFST.

  * Update external links for sfst, foma and flex.

  * Add lex/yacc sources to dist (see issue #524).

* Features:

  * A new variable in the hfst-pmatch language, "unicode-character-classes",
    which causes the use of built-in character sets to be interpreted at
    runtime using a unicode library.

  * Verbose output from hfst-pmatch2fst has been much improved, providing eg.
    a call graph of the process. It also warns about use of pmatch features
    that is likely to be incorrect and lead to errors.


Noteworthy changes in 3.15.5
----------------------------

* Fixes to pmatch:

  * Modify tracking of entry points to allow for RTNs in Captures()

  * Fix off-by-one bug in checking captures

* Fixes to other issues: #508, #488, #486, #452, #491, #494

* Fix memory leaks


Noteworthy changes in 3.15.4
----------------------------

* Import Foma patches with license change to Apache-2.0 and minimal other fixes

* Allow regeneration of Foma files so the generated files can be omitted

* Explicit license file for SFST


Noteworthy changes in 3.15.3
----------------------------

* Numerous improvements to command line tools, e.g. substitute, insert-freely, proc, fst2txt, expand


Noteworthy changes in 3.15.2
----------------------------

* Fix off-by-one bug in checking captures in pmatch


Noteworthy changes in 3.15.1
----------------------------

* Fix issue #443

* Fix issue #429

* Fix issue #409


Noteworthy changes in 3.15.0
----------------------------

* Numerous improvements in pmatch processing:

  * hfst-pmatch2fst now implements eg @include"filename.txt" as a preprocessing stage
  
  * hfst-pmatch2fst now implements Explode() and Implode() in full generality
  
  * hfst-pmatch now allows printing weights (in location mode, with --locate)
  
  * hfst.PmatchContainer.locate offered via python bindings
  
  * bugfixes and improvements hfst.PmatchContainer.tokenize

* Add tool hfst-eliminate-flags.

* Allow both escaped and unescaped zeros in multicharacters in all lexc lexicon entries (fixes issue #274).

* Also reweight state that is both initial and final in hfst-reweight (fixes issue #368).

* Do not let distributed files depend on configuration. Include man files in dist.

* Print warnings and error messages also in earlier stages of twol compilation.

* Add option --execute-and-quit (-E) to hfst-xfst.


Noteworthy changes in 3.14.0
----------------------------

* Numerous improvements to pmatching and tokenization:

  * pmatch now supports the default symbol

  * pmatch now supports reading word embeddings in the binary format


tools/src/hfst-fst2txt.cc
//! @file hfst-fst2txt.cc
//!
//! @brief Transducer array printing command line tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#ifdef WINDOWS
#include <io.h>
#endif


#include <iostream>
#include <fstream>
#include <set>
#include <map>
#include <memory>

using std::set;
using std::map;

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#include <errno.h>

#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "HfstTransducer.h"
#include "HfstInputStream.h"
#include "HfstOutputStream.h"
#include "HfstPrintDot.h"
#include "HfstPrintPCKimmo.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

using hfst::HfstTransducer;
using hfst::HfstInputStream;
using hfst::HfstBasicTransducer;
using hfst::implementations::HfstState;

// add tools-specific variables here
static bool use_numbers=false;
static bool print_weights=false;
static bool do_not_print_weights=false;

enum fst_text_format {
    ATT_TEXT, // AT&T / OpenFst compatible TSV
    DOT_TEXT, // Graphviz / dotty
    PCKIMMO_TEXT, // PCKIMMO format
    PROLOG_TEXT, // prolog format
};

static fst_text_format format=ATT_TEXT;

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE]\n"
        "Print transducer in AT&T, dot, prolog or pckimmo format\n"
        "\n", program_name);

    print_common_program_options(message_out);
    print_common_unary_program_options(message_out);
    fprintf(message_out, "Text format options:\n"
        "  -w, --print-weights          If weights are printed in all cases\n"
        "  -D, --do-not-print-weights   If weights are not printed in any "
        "case\n"
        "  -f, --format=TFMT            Print output in TFMT format "
            "[default=att]\n");
    fprintf(message_out, "\n");
    fprintf(message_out,
          "If OUTFILE or INFILE is missing or -, "
      "standard streams will be used.\n"
          "Unless explicitly requested with option -w or -D, "
      "weights are printed\n"
          "if and only if the transducer is in weighted format.\n"
          "TFMT is one of {att, dot, prolog, pckimmo}.\n"
    );
    fprintf(message_out, "\n");
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
}

int
parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
        {
        HFST_GETOPT_COMMON_LONG,
        HFST_GETOPT_UNARY_LONG,
          // add tool-specific options here
            {"print-weights",        no_argument,       0, 'w'},
            {"do-not-print-weights", no_argument,       0, 'D'},
            {"use-numbers",          no_argument,       0, 'n'},
            {"format",               required_argument, 0, 'f'},
            {0,0,0,0}
        };
        int option_index = 0;
        // add tool-specific options here
        int c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT
                             HFST_GETOPT_UNARY_SHORT "wDnf:",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }

        switch (c)
        {
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-unary.h"
          // add tool-specific cases here
        case 'w':
            print_weights = true;
            break;
    case 'D':
        do_not_print_weights = true;
        break;
    case 'n':
        use_numbers = true;
        break;
    case 'f':
        if ((strcmp(optarg, "att") == 0) || (strcmp(optarg, "AT&T") == 0) ||
            (strcmp(optarg, "openfst") == 0) ||
            (strcmp(optarg, "OpenFst") == 0))
          {
            format = ATT_TEXT;
          }
        else if ((strcmp(optarg, "dot") == 0) ||
            (strcmp(optarg, "graphviz") == 0) ||
            (strcmp(optarg, "GraphViz") == 0))
          {
            format = DOT_TEXT;
          }
        else if (strcmp(optarg, "pckimmo") == 0)
          {
            format = PCKIMMO_TEXT;
          }
        else if ((strcmp(optarg, "prolog") == 0) ||
                 (strcmp(optarg, "Prolog") == 0))
          {
            format = PROLOG_TEXT;
          }
        else
          {
            error(EXIT_FAILURE, 0, "Cannot parse %s as text format; Use one of "
                  "att, pckimmo, dot, prolog", optarg);
          }
        break;
#include "inc/getopt-cases-error.h"
        }
    }

#include "inc/check-params-common.h"
#include "inc/check-params-unary.h"
    return EXIT_CONTINUE;
}


int
process_stream(HfstInputStream& instream, FILE* outf)
{
  //instream.open();
    size_t transducer_n = 0;
    while(instream.is_good())
    {
        transducer_n++;
        HfstTransducer* t=NULL;
        try
          {
            t = new HfstTransducer(instream);
          }
        catch (const TransducerTypeMismatchException &e)
          {
            error(EXIT_FAILURE, 0, "input transducers do not have the same type");
          }
        string inputname = t->get_name();
        if (inputname.length() == 0)
          {
            inputname = inputfilename;
          }
        if (transducer_n == 1)
        {
          verbose_printf("Converting %s...\n", inputname.c_str());
        }
        else
        {
          if (instream.get_type() == hfst::XFSM_TYPE) {
            error(EXIT_FAILURE, 0, "Writing more than one transducer in text format to file not supported for xfsm transducers,\n"
                  "use [hfst-head|hfst-tail|hfst-split] to extract individual transducers from input");
            return EXIT_FAILURE;
          }
          verbose_printf("Converting %s..." SIZE_T_SPECIFIER "\n", inputname.c_str(),
                         transducer_n);
        }

        if(transducer_n > 1)
            fprintf(outf, "--\n");

        bool printw; // whether weights are printed
        hfst::ImplementationType type = t->get_type();
        if (print_weights)
          printw=true;
        else if (do_not_print_weights)
          printw=false;
        else if ( (type == hfst::SFST_TYPE || type == hfst::FOMA_TYPE || type == hfst::XFSM_TYPE) )
          printw = false;
        else if ( (type == hfst::TROPICAL_OPENFST_TYPE || type == hfst::LOG_OPENFST_TYPE) )
          printw = true;
        else  // this should not happen
          printw = true;
    switch (format)
      {
      case ATT_TEXT:
        if (use_numbers) // xfsm case checked earlier
          t->write_in_att_format_number(outf,printw);
        else { // xfsm not yet supported
          //if (type == hfst::XFSM_TYPE) // weights are never printed
          //  t->write_xfsm_transducer_in_att_format(outfilename);
          //else
          t->write_in_att_format(outf,printw);
        }
        break;
      case DOT_TEXT: // xfsm case checked earlier
        fprintf(outf, "// This graph generated with hfst-fst2txt\n");
        hfst::print_dot(outf, *t);
        break;
      case PCKIMMO_TEXT: // xfsm case checked earlier
        hfst::print_pckimmo(outf, *t);
        break;
      case PROLOG_TEXT:
        {
          try
            {
              if (type == hfst::XFSM_TYPE) {
                t->write_xfsm_transducer_in_prolog_format(outfilename); // no name or weights printed
              }
              else {
                std::string namestr = t->get_name();
                std::ostringstream ostr;
                ostr << transducer_n;
                std::string alt_namestr = "NO_NAME_" + ostr.str();

                if (namestr.empty()) {
                  namestr = alt_namestr;
                  if (!silent) {
                    fprintf(stderr, "Transducer has no name, giving it a name '%s'...\n", namestr.c_str()); }
                }
                else {
                  namestr = alt_namestr;
                  if (!silent) {
                    fprintf(stderr, "Renaming transducer into '%s'...\n", namestr.c_str()); }
                }

                t->write_in_prolog_format(outf,namestr,printw);
              }
            }
          catch (const HfstException & e)
            {
              error(EXIT_FAILURE, 0, "Error encountered when writing in prolog format: %s", e.name.c_str());
            }
          break;
        }
      default:
        error(EXIT_FAILURE, 0, "Unknown print format");
      }
    delete t;
    }
    instream.close();
    if (outf != stdout)
      {
        fclose(outf);
      }
    return EXIT_SUCCESS;
}


int main( int argc, char **argv )
{
#ifdef WINDOWS
  _setmode(0, _O_BINARY);
#endif

    hfst_set_program_name(argv[0], "0.3", "HfstFst2Txt");
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE)
    {
        return retval;
    }
    // close buffers, we use streams
    if (inputfile != stdin)
    {
        fclose(inputfile);
    }
    
    verbose_printf("Reading from %s, writing to %s\n",
        inputfilename, outfilename);
    // here starts the buffer handling part
    std::unique_ptr<HfstInputStream> instream;
    try {
      instream.reset((inputfile != stdin) ?
        new HfstInputStream(inputfilename) : new HfstInputStream());
    } catch(const HfstException e)  {
        error(EXIT_FAILURE, 0, "%s is not a valid transducer file",
              inputfilename);
        return EXIT_FAILURE;
    }

    if (instream->get_type() == hfst::XFSM_TYPE)
      {
        if (format == DOT_TEXT) {
          error(EXIT_FAILURE, 0, "Output format 'dot' not supported for xfsm transducers, use 'prolog'");
          return EXIT_FAILURE;
        }
        if (format == PCKIMMO_TEXT) {
          error(EXIT_FAILURE, 0, "Output format 'pckimmo' not supported for xfsm transducers, use 'prolog'");
          return EXIT_FAILURE;
        }
        if (format == ATT_TEXT) {
          error(EXIT_FAILURE, 0, "Output format 'att' not supported for xfsm transducers, use 'prolog'");
          return EXIT_FAILURE;
        }
        if (use_numbers) {
          error(EXIT_FAILURE, 0, "Option '--use-numbers' not supported for xfsm transducers");
          return EXIT_FAILURE;
        }
        if (strcmp(inputfilename, "<stdin>") == 0) {
          error(EXIT_FAILURE, 0, "Reading from standard input not supported for xfsm transducers,\n"
                "use 'hfst-fst2txt [--input|-i] INFILE' instead");
          return EXIT_FAILURE;
        }
        if (strcmp(outfilename, "<stdout>") == 0) {
          error(EXIT_FAILURE, 0, "Writing to standard output not supported for xfsm transducers,\n"
                "use 'hfst-fst2txt [--output|-o] OUTFILE' instead");
          return EXIT_FAILURE;
        }
      }
    
    retval = process_stream(*instream, outfile);

    free(inputfilename);
    free(outfilename);
    return retval;
}

